File: app/src/main/java/com/todobom/opennotescanner/helpers/AnimationRunnable.java
package com.todobom.opennotescanner.helpers;

import android.graphics.Bitmap;
import android.util.Log;
import android.view.Display;
import android.view.View;
import android.view.animation.AccelerateInterpolator;
import android.view.animation.Animation;
import android.view.animation.AnimationSet;
import android.view.animation.ScaleAnimation;
import android.view.animation.TranslateAnimation;
import android.widget.ImageView;
import android.widget.RelativeLayout;

import com.todobom.opennotescanner.OpenNoteScannerActivity;
import com.todobom.opennotescanner.R;

import org.opencv.core.Point;
import org.opencv.core.Size;

import static com.todobom.opennotescanner.helpers.Utils.decodeSampledBitmapFromUri;

public class AnimationRunnable implements Runnable {

	private static final String TAG = "AnimationRunnable";

	private final OpenNoteScannerActivity activity;
	private Size imageSize;
	private Point[] previewPoints = null;
	public Size previewSize = null;
	public String fileName = null;
	public int width;
	public int height;
	private Bitmap bitmap;

	public AnimationRunnable(OpenNoteScannerActivity activity, String filename, ScannedDocument document) {
		this.activity = activity;
		this.fileName = filename;
		this.imageSize = document.processed.size();

		if (document.quadrilateral != null) {
			this.previewPoints = document.previewPoints;
			this.previewSize = document.previewSize;
		}
	}

	public double hipotenuse(Point a, Point b) {
		return Math.sqrt(Math.pow(a.x - b.x, 2) + Math.pow(a.y - b.y, 2));
	}

	@Override
	public void run() {
		final ImageView imageView = (ImageView) activity.findViewById(R.id.scannedAnimation);

		Display display = activity.getWindowManager().getDefaultDisplay();
		android.graphics.Point size = new android.graphics.Point();
		display.getRealSize(size);

		int width = Math.min(size.x, size.y);
		int height = Math.max(size.x, size.y);

		// ATENTION: captured images are always in landscape, values should be swapped
		double imageWidth = imageSize.height;
		double imageHeight = imageSize.width;

		RelativeLayout.LayoutParams params = (RelativeLayout.LayoutParams) imageView.getLayoutParams();

		if (previewPoints != null) {
			double documentLeftHeight = hipotenuse(previewPoints[0], previewPoints[1]);
			double documentBottomWidth = hipotenuse(previewPoints[1], previewPoints[2]);
			double documentRightHeight = hipotenuse(previewPoints[2], previewPoints[3]);
			double documentTopWidth = hipotenuse(previewPoints[3], previewPoints[0]);

			double documentWidth = Math.max(documentTopWidth, documentBottomWidth);
			double documentHeight = Math.max(documentLeftHeight, documentRightHeight);

			Log.d(TAG, "device: " + width + "x" + height + " image: " + imageWidth + "x" + imageHeight + " document: " + documentWidth + "x" + documentHeight);


			Log.d(TAG, "previewPoints[0] x=" + previewPoints[0].x + " y=" + previewPoints[0].y);
			Log.d(TAG, "previewPoints[1] x=" + previewPoints[1].x + " y=" + previewPoints[1].y);
			Log.d(TAG, "previewPoints[2] x=" + previewPoints[2].x + " y=" + previewPoints[2].y);
			Log.d(TAG, "previewPoints[3] x=" + previewPoints[3].x + " y=" + previewPoints[3].y);

			// ATENTION: again, swap width and height
			double xRatio = width / previewSize.height;
			double yRatio = height / previewSize.width;

			params.topMargin = (int) (previewPoints[3].x * yRatio);
			params.leftMargin = (int) ((previewSize.height - previewPoints[3].y) * xRatio);
			params.width = (int) (documentWidth * xRatio);
			params.height = (int) (documentHeight * yRatio);
		} else {
			params.topMargin = height / 4;
			params.leftMargin = width / 4;
			params.width = width / 2;
			params.height = height / 2;
		}

		bitmap = decodeSampledBitmapFromUri(fileName, params.width, params.height);

		imageView.setImageBitmap(bitmap);

		imageView.setVisibility(View.VISIBLE);

		TranslateAnimation translateAnimation = new TranslateAnimation(
				Animation.ABSOLUTE, 0, Animation.ABSOLUTE, -params.leftMargin,
				Animation.ABSOLUTE, 0, Animation.ABSOLUTE, height - params.topMargin
		);

		ScaleAnimation scaleAnimation = new ScaleAnimation(1, 0, 1, 0);

		AnimationSet animationSet = new AnimationSet(true);

		animationSet.addAnimation(scaleAnimation);
		animationSet.addAnimation(translateAnimation);

		animationSet.setDuration(600);
		animationSet.setInterpolator(new AccelerateInterpolator());

		animationSet.setAnimationListener(new Animation.AnimationListener() {
			@Override
			public void onAnimationStart(Animation animation) {

			}

			@Override
			public void onAnimationEnd(Animation animation) {
				imageView.setVisibility(View.INVISIBLE);
				imageView.setImageBitmap(null);
				if (AnimationRunnable.this.bitmap != null) {
					AnimationRunnable.this.bitmap.recycle();
				}
			}

			@Override
			public void onAnimationRepeat(Animation animation) {

			}
		});


		imageView.startAnimation(animationSet);

	}
}

File: app/src/main/java/com/todobom/opennotescanner/helpers/AnimationRunnable.kt
package com.todobom.opennotescanner.helpers

import android.graphics.Bitmap
import android.util.Log
import android.view.View
import android.view.animation.*
import android.widget.ImageView
import android.widget.RelativeLayout
import com.todobom.opennotescanner.OpenNoteScannerActivity
import com.todobom.opennotescanner.R
import org.opencv.core.Point
import org.opencv.core.Size

class AnimationRunnable(private val activity: OpenNoteScannerActivity, filename: String, document: ScannedDocument) : Runnable {
    private val imageSize: Size = document.processed!!.size()
    private val previewPoints: Array<Point>? = if(document.quadrilateral != null) document.previewPoints else null
    private val previewSize: Size? = if(document.quadrilateral != null) document.previewSize else null
    private val fileName: String = filename
    private var bitmap: Bitmap? = null

    override fun run() {
        val imageView = activity.findViewById<View>(R.id.scannedAnimation) as ImageView
        val display = activity.windowManager.defaultDisplay
        val size = android.graphics.Point()
        display.getRealSize(size)
        val width = Math.min(size.x, size.y)
        val height = Math.max(size.x, size.y)

        // ATENTION: captured images are always in landscape, values should be swapped
        val imageWidth = imageSize.height
        val imageHeight = imageSize.width
        val params = imageView.layoutParams as RelativeLayout.LayoutParams
        val prevPoints = previewPoints
        val prevSize = previewSize

        if (prevPoints != null) {
            prevSize!!

            val documentLeftHeight = hipotenuse(prevPoints[0], prevPoints[1])
            val documentBottomWidth = hipotenuse(prevPoints[1], prevPoints[2])
            val documentRightHeight = hipotenuse(prevPoints[2], prevPoints[3])
            val documentTopWidth = hipotenuse(prevPoints[3], prevPoints[0])
            val documentWidth = Math.max(documentTopWidth, documentBottomWidth)
            val documentHeight = Math.max(documentLeftHeight, documentRightHeight)
            Log.d(TAG, "device: " + width + "x" + height + " image: " + imageWidth + "x" + imageHeight + " document: " + documentWidth + "x" + documentHeight)
            Log.d(TAG, "prevPoints[0] x=" + prevPoints[0].x + " y=" + prevPoints[0].y)
            Log.d(TAG, "prevPoints[1] x=" + prevPoints[1].x + " y=" + prevPoints[1].y)
            Log.d(TAG, "prevPoints[2] x=" + prevPoints[2].x + " y=" + prevPoints[2].y)
            Log.d(TAG, "prevPoints[3] x=" + prevPoints[3].x + " y=" + prevPoints[3].y)

            // ATENTION: again, swap width and height
            val xRatio = width / prevSize.height
            val yRatio = height / prevSize.width
            params.topMargin = (prevPoints[3].x * yRatio).toInt()
            params.leftMargin = ((prevSize.height - prevPoints[3].y) * xRatio).toInt()
            params.width = (documentWidth * xRatio).toInt()
            params.height = (documentHeight * yRatio).toInt()
        } else {
            params.topMargin = height / 4
            params.leftMargin = width / 4
            params.width = width / 2
            params.height = height / 2
        }

        bitmap = Utils.decodeSampledBitmapFromUri(fileName, params.width, params.height)
        imageView.setImageBitmap(bitmap)
        imageView.visibility = View.VISIBLE
        val translateAnimation = TranslateAnimation(
                Animation.ABSOLUTE, 0f, Animation.ABSOLUTE, (-params.leftMargin).toFloat(),
                Animation.ABSOLUTE, 0f, Animation.ABSOLUTE, (height - params.topMargin).toFloat()
        )
        val scaleAnimation = ScaleAnimation(1f, 0f, 1f, 0f)
        val animationSet = AnimationSet(true)
        animationSet.addAnimation(scaleAnimation)
        animationSet.addAnimation(translateAnimation)
        animationSet.duration = 600
        animationSet.interpolator = AccelerateInterpolator()
        animationSet.setAnimationListener(object : Animation.AnimationListener {
            override fun onAnimationStart(animation: Animation) {}
            override fun onAnimationEnd(animation: Animation) {
                imageView.visibility = View.INVISIBLE
                imageView.setImageBitmap(null)
                bitmap?.recycle()
            }

            override fun onAnimationRepeat(animation: Animation) {}
        })
        imageView.startAnimation(animationSet)
    }

    companion object {
        private const val TAG = "AnimationRunnable"

        @JvmStatic
        fun hipotenuse(a: Point, b: Point): Double {
            return Math.sqrt(Math.pow(a.x - b.x, 2.0) + Math.pow(a.y - b.y, 2.0))
        }
    }
}